Faster proximal learning in column pooler

mrcslws opened the pull request:
This is for RES-357.

This makes the column pooler about 50% faster, according to `projects/l2_pooling/multi_column_convergence.py`. Running 

~~~python
runExperiment({
  "numObjects": 10,
  "numLocations": 5,
  "numFeatures": 5,
  "numColumns": 7,
  "trialNum": 0
  })
~~~

previously took 19.6 seconds, and now takes 13.1 seconds.

subutai commented:
@mrcslws Nice!  Do the tests in tests/algorithms and tests/layers continue to pass?

subutai commented:
Also tests/regions/column_pooler_region_test.py

subutai commented on htmresearch/algorithms/column_pooler.py:
```diff
@@ -532,3 +476,22 @@ def _winnersBasedOnLateralActivity(self,
     return predictedActiveCells
+
+
+def sample(iterable, k, rng):
```
Should this be _sample?

mrcslws commented:
@subutai I replaced the `testPickProximalInputsToLearnOn` since the `_pickProximalInputsToLearnOn` method no longer exists. These new tests cover `_learnProximal` as a whole. They're still unit tests, but they test the units via the `compute` method.

(I've had success with this approach in the TemporalMemory unit tests. Since I added compute-based unit tests to the TM, I've changed the TM many times and the unit tests just keep working. Writing the unit tests this way is more difficult, but the tests survive much longer and they make development more rapid.)

It passes all tests now.

subutai commented:
👍 Thanks - these look good!
